picard/script.py

```diff
@@ -145,13 +174,13 @@ def eval(self, parser):
         try:
             function, eval_args, num_args = parser.functions[self.name]
         except KeyError:
-            raise ScriptUnknownFunction("Unknown function '%s'" % self.name)
+            raise ScriptUnknownFunction("Unknown function '%s'" % self.name, self.stackitem)
```

Comment:
I wonder if we could just pass a special stackitem:

`UnknownFunctionStackItem(self.stackitem.line, self.stackitem.column, self.name)`

or even:

`UnknownFunctionStackItem(self.stackitem, self.name)`

 and format the message in `ScriptUnknownFunction()`.

Here what I'm thinking about:

```python
class StackItem:
    def __init__(self, line, column):
        self.line = line
        self.column = column

    def __str__(self):
        return '{line:d}:{column:d}'.format(
            line=self.line,
            column=self.column
        )


class FunctionStackItem(StackItem):
    def __init__(self, line, column, name):
        super().__init__(line, column)
        self.name = '$' + name

    def __str__(self):
        return '{prefix:s}:{name:s}'.format(
            prefix=super().__str__(),
            name=self.name
        )


class UnknownFunctionStackItem(FunctionStackItem):
    def __init__(self, *args):
        if isinstance(args[0], StackItem):
            stackitem = args[0]
            line = stackitem.line
            column = stackitem.column
            name = args[1]
        else:
            line, column, name = args
        super().__init__(line, column, name)

    def __str__(self):
        return '{prefix:s} Unknown function'.format(
            prefix=super().__str__()
        )
```

```python
stackitem = StackItem(4, 8)
print(str(UnknownFunctionStackItem(stackitem, "xxx")))
print(str(UnknownFunctionStackItem(9, 54, "yyy")))
```

outputs:

```
4:8:$xxx Unknown function
9:54:$yyy Unknown function
```


Good or bad idea?

Reply:
Actually we don't need different classes (suggested by @phw).

```python
class StackItem:
    def __init__(self, line, column, name=None):
        self.line = line
        self.column = column
        self.name = name

    def __str__(self):
        if name is None:
            return '{line:d}:{column:d}'.format(
                line=self.line,
                column=self.column
            )
        else:
            return '{line:d}:{column:d}:${name}'.format(
                line=self.line,
                column=self.column,
                name=self.name
            )
```

Reply:
I used the combined class as suggested.  (Great idea.)